# Week 3 Assignment: Heart Disease Dataset Analysis

## I 320D: Data Science for Biomedical Informatics | Spring 2026

### 📋 Assignment Version C

---

## 🎯 This Week's Mantra

> **"Every Column Tells a Story"**

In this assignment, you'll apply the 10-Point Data Inspection to a real-world healthcare dataset focused on heart disease diagnosis. By the end, you'll understand not just *what* the data contains, but *why* each variable matters for clinical decision-making.

---

## Learning Objectives

By completing this assignment, you will be able to:

1. ✅ Apply the systematic 10-Point Inspection to a new healthcare dataset
2. ✅ Identify and classify feature types (continuous, discrete, categorical, ordinal)
3. ✅ Detect and document data quality issues (missing values, unexpected values)
4. ✅ Research and document clinical meaning for healthcare variables
5. ✅ Create meaningful data groupings based on clinical standards
6. ✅ Formulate answerable research questions about heart disease risk factors

---

## About the Dataset

**Dataset:** Heart Disease UCI (Combined)  
**Source:** UCI Machine Learning Repository / Kaggle  
**File:** `heart_disease_uci.csv`  
**Target Variable:** `num` (diagnosis of heart disease: 0 = no disease, 1-4 = presence of disease)

### Clinical Context

Heart disease remains the leading cause of death globally, accounting for approximately 17.9 million deaths annually according to the World Health Organization. This dataset combines patient records from four medical institutions: Cleveland Clinic, Hungarian Institute of Cardiology, University Hospital Zurich (Switzerland), and VA Long Beach Medical Center. Understanding these diagnostic variables is crucial for:

- Early identification of high-risk patients
- Understanding regional variations in heart disease presentation
- Clinical decision support systems
- Risk stratification and preventive care planning

---

## Getting Started

First, load the dataset and import your libraries:

In [4]:
# Import libraries
import pandas as pd
import numpy as np

# Load the dataset
df = pd.read_csv('heart_disease_uci.csv')

# Display first few rows to confirm it loaded
print("\nFirst 5 Rows:")
df.head()


First 5 Rows:


,id,age,sex,dataset,cp,trestbps,chol,fbs,restecg,thalch,exang,oldpeak,slope,ca,thal,num
0,1,63,Male,Cleveland,typical angina,145.0,233.0,True,lv hypertrophy,150.0,False,2.3,downsloping,0.0,fixed defect,0
1,2,67,Male,Cleveland,asymptomatic,160.0,286.0,False,lv hypertrophy,108.0,True,1.5,flat,3.0,normal,2
2,3,67,Male,Cleveland,asymptomatic,120.0,229.0,False,lv hypertrophy,129.0,True,2.6,flat,2.0,reversable defect,1
3,4,37,Male,Cleveland,non-anginal,130.0,250.0,False,normal,187.0,False,3.5,downsloping,0.0,normal,0
4,5,41,Female,Cleveland,atypical angina,130.0,204.0,False,lv hypertrophy,172.0,False,1.4,upsloping,0.0,normal,0


---

## Part 1: The 10-Point Data Inspection (40 points)

Complete each inspection step and document your findings.

### Step 1: Shape (4 points)

**Your Code:**

In [20]:
print("Shape:", df.shape)


Shape: (920, 16)


**Your Findings:**
- How many rows (observations)?

   920
- How many columns (features)?
  
    16
- What does each row represent in clinical terms?
  
  Each row represents one individual patient’s clinical record.

---

### Step 2: Column Names (4 points)

**Your Code:**

In [9]:
print("Column names:", df.columns)

Column names: Index(['id', 'age', 'sex', 'dataset', 'cp', 'trestbps', 'chol', 'fbs',
       'restecg', 'thalch', 'exang', 'oldpeak', 'slope', 'ca', 'thal', 'num'],
      dtype='str')


**Your Findings:**
- List all column names:

    'id', 'age', 'sex', 'dataset', 'cp', 'trestbps', 'chol', 'fbs',
       'restecg', 'thalch', 'exang', 'oldpeak', 'slope', 'ca', 'thal', 'num'

- Which columns might need further research to understand? (Hint: Many use medical abbreviations!):

    'cp', 'trestbps', 'chol', 'fbs',
       'restecg', 'thalch', 'exang', 'oldpeak', 'slope', 'ca', 'thal'

---

### Step 3: Data Types (4 points)

**Your Code:**

In [10]:
print("Data Types:", df.dtypes)

Data Types: id            int64
age           int64
sex             str
dataset         str
cp              str
trestbps    float64
chol        float64
fbs          object
restecg         str
thalch      float64
exang        object
oldpeak     float64
slope           str
ca          float64
thal            str
num           int64
dtype: object


**Your Findings:**
- Which columns are numeric (int64 or float64)?

    id, age, trestbps, chol, thalch, pldpeak, ca, num

- Which columns are categorical (object/string)?

   sex, dataset, cp, fbs, restecg, exang, slope, thal 

- Are there any data types that seem incorrect based on what you know about the data?

    No

---

### Step 4: First Look (4 points)

**Your Code:**

In [11]:
print("First Look", df.head())

First Look    id  age     sex    dataset               cp  trestbps   chol    fbs  \
0   1   63    Male  Cleveland   typical angina     145.0  233.0   True   
1   2   67    Male  Cleveland     asymptomatic     160.0  286.0  False   
2   3   67    Male  Cleveland     asymptomatic     120.0  229.0  False   
3   4   37    Male  Cleveland      non-anginal     130.0  250.0  False   
4   5   41  Female  Cleveland  atypical angina     130.0  204.0  False   

          restecg  thalch  exang  oldpeak        slope   ca  \
0  lv hypertrophy   150.0  False      2.3  downsloping  0.0   
1  lv hypertrophy   108.0   True      1.5         flat  3.0   
2  lv hypertrophy   129.0   True      2.6         flat  2.0   
3          normal   187.0  False      3.5  downsloping  0.0   
4  lv hypertrophy   172.0  False      1.4    upsloping  0.0   

                thal  num  
0       fixed defect    0  
1             normal    2  
2  reversable defect    1  
3             normal    0  
4             normal    0

**Your Findings:**
- What do the actual values look like?

  The actual value contain numeric clinical data, binary indicators and text-based clinical categories.

- Do you notice any categorical variables that are already human-readable vs. encoded?

  Yes. Many categorical variables are already human-readable. This improves interpretability and reduces the need for decoding.

- Are there any values that look like they might be placeholders for missing data?

    Yes. Some columns in the full dataset contain values of 0.

  
---

### Step 5: Last Look (4 points)

**Your Code:**

In [12]:
print("Last Look", df.tail())

Last Look       id  age     sex        dataset               cp  trestbps   chol    fbs  \
915  916   54  Female  VA Long Beach     asymptomatic     127.0  333.0   True   
916  917   62    Male  VA Long Beach   typical angina       NaN  139.0  False   
917  918   55    Male  VA Long Beach     asymptomatic     122.0  223.0   True   
918  919   58    Male  VA Long Beach     asymptomatic       NaN  385.0   True   
919  920   62    Male  VA Long Beach  atypical angina     120.0  254.0  False   

              restecg  thalch  exang  oldpeak slope  ca          thal  num  
915  st-t abnormality   154.0  False      0.0   NaN NaN           NaN    1  
916  st-t abnormality     NaN    NaN      NaN   NaN NaN           NaN    0  
917  st-t abnormality   100.0  False      0.0   NaN NaN  fixed defect    2  
918    lv hypertrophy     NaN    NaN      NaN   NaN NaN           NaN    0  
919    lv hypertrophy    93.0   True      0.0   NaN NaN           NaN    1  


**Your Findings:**
- Does the data end cleanly?

    Yes. There are no formatting errors, truncated rows, or structural problems at the end of the file.
- Are the last rows consistent with the first rows?

    Yes. In structure, but not in data.

The same 16 columns are present.
- Do you notice more or fewer missing values in later rows?
  
    Yes, there are more missing values in later rows, particularly in: thal, thalch, cp, slope, oldpeak, ca

---

### Step 6: Memory Usage (4 points)

**Your Code:**

In [13]:
print("Memory Usage", df.memory_usage())

Memory Usage Index        132
id          7360
age         7360
sex         7360
dataset     7360
cp          7360
trestbps    7360
chol        7360
fbs         7360
restecg     7360
thalch      7360
exang       7360
oldpeak     7360
slope       7360
ca          7360
thal        7360
num         7360
dtype: int64


**Your Findings:**
- How much memory does the dataset use? _______________ MB

In [60]:
total_memory_mb = df.memory_usage(deep=True).sum() / (1024**2)

print("Total Memory Usage (MB):", round(total_memory_mb, 4))


Total Memory Usage (MB): 0.4626


- Is this a "small" or "large" dataset by data science standards?

  This is a small dataset by data science standards.


---

### Step 7: Missing Values (4 points)

**Your Code:**

In [14]:
print("Missing Value", df.isnull().sum())

Missing Value id            0
age           0
sex           0
dataset       0
cp            0
trestbps     59
chol         30
fbs          90
restecg       2
thalch       55
exang        55
oldpeak      62
slope       309
ca          611
thal        486
num           0
dtype: int64


**Your Findings:**
- Which columns have missing values?

  Trestbps, chol, fbs, restecg, thalch, exang, oldpeak, slope, ca, and thal.

- What percentage of each column is missing? 

In [62]:
missing_counts = df.isnull().sum()

missing_percent = (df.isnull().mean() * 100)

missing_summary = pd.DataFrame({
    "Missing Count": missing_counts,
    "Missing Percent (%)": missing_percent.round(2)
})

missing_summary = missing_summary[missing_summary["Missing Count"] > 0]

print(missing_summary.sort_values("Missing Percent (%)", ascending=False))



          Missing Count  Missing Percent (%)
ca                  611                66.41
thal                486                52.83
slope               309                33.59
chol                202                21.96
fbs                  90                 9.78
oldpeak              62                 6.74
trestbps             59                 6.41
thalch               55                 5.98
exang                55                 5.98
restecg               2                 0.22


- Which columns have the MOST missing data? What might explain this?

In [65]:
print("The MOST missing data:", missing_summary.max())

The MOST missing data: Missing Count          611.00
Missing Percent (%)     66.41
dtype: float64


---

### Step 8: Duplicates (4 points)

**Your Code:**

In [15]:
print("Duplicates", df.duplicated())

Duplicates 0      False
1      False
2      False
3      False
4      False
       ...  
915    False
916    False
917    False
918    False
919    False
Length: 920, dtype: bool


**Your Findings:**
- Are there any duplicate rows? _____NO__________
- Are all patient IDs unique? ________YES_______

---

### Step 9: Basic Statistics (4 points)

**Your Code:**

In [18]:
print("Basic Statistic")
print(df.describe())

Basic Statistic
               id         age    trestbps        chol      thalch     oldpeak  \
count  920.000000  920.000000  861.000000  890.000000  865.000000  858.000000   
mean   460.500000   53.510870  132.132404  199.130337  137.545665    0.878788   
std    265.725422    9.424685   19.066070  110.780810   25.926276    1.091226   
min      1.000000   28.000000    0.000000    0.000000   60.000000   -2.600000   
25%    230.750000   47.000000  120.000000  175.000000  120.000000    0.000000   
50%    460.500000   54.000000  130.000000  223.000000  140.000000    0.500000   
75%    690.250000   60.000000  140.000000  268.000000  157.000000    1.500000   
max    920.000000   77.000000  200.000000  603.000000  202.000000    6.200000   

               ca         num  
count  309.000000  920.000000  
mean     0.676375    0.995652  
std      0.935653    1.142693  
min      0.000000    0.000000  
25%      0.000000    0.000000  
50%      0.000000    1.000000  
75%      1.000000    2.000000 

## **Your Findings:**
- What is the age range in the dataset? ____28___________ to ____77___________year
- What is the range of resting blood pressure (trestbps)? _____0__________ to __200_____________mmHG
- What is the range of serum cholesterol (chol)? ________0_______ to _______603________mg/dL
- What is the range of maximum heart rate achieved (thalch)? _______60________ to ____202___________bpm
- Do any min/max values seem impossible or clinically unlikely?

  Yes. Minimum oldpeak, trestbps and chol are impossible.
  
---

### Step 10: Unique Counts (4 points)

**Your Code:**

In [59]:
print("Unique Count")
print(df.nunique())

Unique Count
id                920
age                50
sex                 2
dataset             4
cp                  4
trestbps           61
chol              216
fbs                 2
restecg             3
thalch            119
exang               2
oldpeak            53
slope               3
ca                  4
thal                3
num                 5
chol_category       4
disease_binary      2
dtype: int64


**Your Findings:**
- Which columns have very few unique values (likely categorical)?

  Sex, dataset, cp, fbs, restecg, exang, slope, ca, thal, num, chol_category and disease_binary.

- Which columns have many unique values (likely continuous or IDs)?

  Id, age, trestbps, chol, thalch, oldpeak.

- Does the number of unique IDs match the number of rows? _____YES__________

---

## Part 2: Data Dictionary (20 points)

Complete the following data dictionary. For each column, you must:
1. **Research** the clinical meaning (these are standard cardiac assessment terms)
2. **Identify** the feature type (Continuous, Discrete, Categorical-Nominal, Categorical-Ordinal, Binary, Identifier)
3. **Document** the valid values/range you observe
4. **Note** any issues or questions

| Column | Description | Feature Type | Valid Values/Range | Notes/Issues |
|--------|-------------|--------------|-------------------|--------------|
| `id` | `id Unique patient identifier` | `Identififier` | `integer` | `no clinical meaning` |
| `age` | `patients' age` | `continous` | `29-77` | `integer` |
| `sex` | `sex of patient` | `binary` | `male, female` | `NA` |
| `dataset` | `Source medical center` | `Nominal` | `Cleveland, Hungary, Switzerland, VA Long Beach`| |
| `cp` | `Chest pain type` | `Categorical-Nominal` | `Typical angina, Atypical angina, Non-anginal pain, Asymptomatic` | `Strong predictor of heart disease` |
| `trestbps` | `Resting blood pressure` | `continous` | `0–~200+` | `0 is not physiologically possible` |
| `chol` | `Serum total cholesterol (mg/dL)` | `continous` | `0–~600` | `0 is not physiologically possible` |
| `fbs` | `Fasting blood sugar >120 mg/dL` | `Binary` | `TRUE / FALSE` | `Indicator of diabetes risk` |
| `restecg` | `Resting ECG result` | `Categorical-Nominal` | `Normal, ST-T abnormality, LV hypertrophy` | `Based on electrocardiogram` |
| `thalch` | `Maximum heart rate achieved` | `continous` | `0–202 bpm` | |
| `exang` | `Exercise-induced angina` | `Binary` | `TRUE / FALSE` | |
| `oldpeak` | `ST depression induced by exercise` | `continous` | `0.0–~6.0` | `Measured in mm`|
| `slope` | `Slope of peak exercise ST segment` | `Categorical-Nominal` | `Upsloping, Flat, Downslop` | |
| `ca` | `Number of major vessels colored by fluoroscopy` | `Discrete` | `0–3 (sometimes 4)` | |
| `thal` | `Thalassemia test result` | `Categorical-Nominal` | `Normal, Fixed defect, Reversible defect` | |
| `num` | `Diagnosis of heart disease severity` | `Discrete` | `0–4` | |

### Clinical Research Questions for Version C

Answer these questions based on your research (you may need to use Google):

**1. What is serum cholesterol? According to current guidelines, what are the recommended total cholesterol levels (desirable, borderline high, and high)?**

Your answer:
Serum cholesterol is the total cholesterol level in the blood. It includes LDL (bad), HDL (good), and VLDL cholesterol.

Desirable:<200

Borderline High:200-239

High:>=240

---

**2. What is maximum heart rate achieved (thalch) during a stress test? How is target heart rate calculated, and why is achieving a certain percentage of maximum heart rate important for test validity?**

Your answer:
The maximum heart rate during a stress test is the highest beats per minute (BPM) achieved under exertion, generally calculated as 200-age.

Stress tests are considered valid if a patient reaches 85% of predicted maximum heart rate.

Why this matters:

If patient does not reach adequate heart rate, the test may not detect ischemia.

Low achieved heart rate may indicate cardiac dysfunction.

---

**3. What is the ST segment slope (slope) during exercise? What do "upsloping," "flat," and "downsloping" patterns indicate about heart health?**

Your answer:
The ST segment slope refers to the pattern of the ST segment on ECG during peak exercise.

Types:

Upsloping: normal or benign

Flat: high-risk finding for heart disease(typically >= 0.5 mm or 1 mm, depending on criteria)

Downsloping: Strongly significant myocardial ischemia 

---

**4. Why might this dataset combine data from four different countries/institutions? What are the potential benefits and challenges of multi-center data for heart disease research?**

Your answer:
The dataset combines multiple countries can improve model reliability, reflect patient diversity, increase sample size and so on.

Benefit:

More diverse patient populations

Better machine learning training data

Challenges:

Different measurement protocols

Possible demographic differences

---

## Part 3: Data Validation (15 points)

### 3.1 Blood Pressure Validation (5 points)

Write code to check:
- How many blood pressure readings are 0? (This is clinically impossible for a living patient!)
- How many readings are below 80 mmHg? Above 200 mmHg?
- What might a value of 0 represent in this dataset?

**Your Code:**

In [24]:
bp_zero = (df["trestbps"] == 0).sum()
bp_below_80 = (df["trestbps"] < 80).sum()
bp_above_200 = (df["trestbps"] > 200).sum()



**Your Findings:**

- Are there any impossible blood pressure values?

- How should values of 0 be treated - as missing data or as valid values?

- Does this match what `.isnull()` reports for this column?

In [25]:
print("There are", bp_zero + bp_below_80 + bp_above_200, "impossible blood pressure value.")


There are 2 impossible blood pressure value.


Any values of 0 likely represent missing data, not real measurements.

.isnull() originally does not detect zeros, so manual cleaning is required.

---

### 3.2 Cholesterol Validation (5 points)

Write code to examine cholesterol values closely. Check for any unusual values including zeros.

**Your Code:**

In [26]:
# Count zero cholesterol values
chol_zero = (df["chol"] == 0).sum()

# Check summary statistics
df["chol"].describe()


count    890.000000
mean     199.130337
std      110.780810
min        0.000000
25%      175.000000
50%      223.000000
75%      268.000000
max      603.000000
Name: chol, dtype: float64

**Your Findings:**

- Did you find any cholesterol values of 0?

  Yes, the minimum value is 0.

- Is a cholesterol level of 0 clinically possible?

  No. A cholesterol level of 0 mg/dL is physiologically impossible in a living human. Cholesterol is required for cell membranes and hormone production.

- How many such impossible values exist?

In [28]:
print("There are", chol_zero, "impossible value.")


There are 172 impossible value.


- What would happen if you calculated the mean cholesterol without handling this?
  The mean cholesterol may be lowered.

---

### 3.3 Multi-Source Data Validation (5 points)

This dataset combines data from four different medical centers. Write code to:
1. Count patients from each source (dataset column)
2. Check if missing data patterns differ by source

**Your Code:**

In [37]:
source_counts = df["dataset"].value_counts()
print(source_counts)

dataset
Cleveland        304
Hungary          293
VA Long Beach    200
Switzerland      123
Name: count, dtype: int64


**Your Findings:**

- Which medical center contributed the most patients?

In [51]:
print("Center with MOST patients:", source_counts.idxmax())

Center with MOST patients: Cleveland


- Which medical center contributed the fewest?

In [52]:
print("Center with FEWEST patients:", source_counts.idxmin())

Center with FEWEST patients: Switzerland


- Do certain columns have more missing data in certain sources? Which ones?
  
  Yes. Missing data varies by institution.
  

- Why might data completeness vary across medical institutions?
Reason:
Different diagnostic protocols
Different equipment availability
Some tests not routinely performed at all centers

---

## Part 4: Create Clinical Cholesterol Groups (10 points)

Create a new column called `chol_category` that categorizes patients based on their serum cholesterol levels according to clinical guidelines.

### Version C: Cholesterol Risk Categories

Use these categories based on clinical cholesterol guidelines:

| Cholesterol Category | Range (mg/dL) | Clinical Significance |
|---------------------|---------------|----------------------|
| Missing/Invalid | 0 or NaN | Data quality issue - cannot classify |
| Desirable | < 200 | Healthy cholesterol level |
| Borderline High | 200-239 | Increased cardiovascular risk |
| High | ≥ 240 | High cardiovascular risk - treatment recommended |

### Your Code:

In [43]:
# Create the chol_category column
def classify_chol(chol):
    if pd.isna(chol):
        return "Missing/Invalid"
    elif chol < 200:
        return "Desirable"
    elif 200 <= chol <= 239:
        return "Borderline High"
    else:
        return "High"

df["chol_category"] = df["chol"].apply(classify_chol)

# IMPORTANT: You must handle values of 0 appropriately - they should be "Missing/Invalid"
df.loc[df["chol"] == 0, "chol"] = np.nan





### Verify your groupings worked:

In [44]:
# Show counts per cholesterol category
print("Counts per cholesterol category:")
print(df["chol_category"].value_counts())



Counts per cholesterol category:
chol_category
High               359
Borderline High    231
Missing/Invalid    202
Desirable          128
Name: count, dtype: int64


### Calculate heart disease rate by cholesterol category:

In [46]:
# Create a binary indicator for heart disease (num > 0 means disease present)
df["disease_binary"] = np.where(df["num"] > 0, 1, 0)
# Exclude the "Missing/Invalid" category from your interpretation
valid_df = df[df["chol_category"] != "Missing/Invalid"]
# Then calculate the percentage of patients with heart disease in each cholesterol category
disease_rate = valid_df.groupby("chol_category")["disease_binary"].mean() * 100

print("Heart disease rate by cholesterol category (%):\n")
print(disease_rate.round(2))




Heart disease rate by cholesterol category (%):

chol_category
Borderline High    46.32
Desirable          37.50
High               53.20
Name: disease_binary, dtype: float64


### Analysis Questions:

**1. How many patients are in each cholesterol category? How many have missing/invalid cholesterol readings?**

Your answer:

In [47]:
print(df["chol_category"].value_counts())

chol_category
High               359
Borderline High    231
Missing/Invalid    202
Desirable          128
Name: count, dtype: int64


---

**2. What is the heart disease rate (percentage) for each cholesterol category (excluding invalid)?**

Your answer:

In [48]:
print(disease_rate.round(2))

chol_category
Borderline High    46.32
Desirable          37.50
High               53.20
Name: disease_binary, dtype: float64


---

**3. Does heart disease prevalence increase with cholesterol level as expected? If the pattern is surprising, what factors might explain it?**

Your answer:
Yes. The pattern shows a steady increase:

37.50% (Desirable)

46.32% (Borderline High)

53.20% (High)

Heart disease prevalence increases as cholesterol level increases. This matches clinical expectations. The increase is gradual rather than extreme.

---

## Part 5: Research Questions (15 points)

### 5.1 Write Three Answerable Questions (9 points)

Write three questions that THIS dataset can answer. Remember: the data can show relationships and patterns, but cannot prove causation.

**Your questions must explore these specific areas:**

1. **A question about ST slope (slope) and heart disease:**

   The increase is gradual rather than extreme.

---

2. **A question comparing patients with vs. without exercise-induced angina (exang):**

    Do patients with exercise-induced angina have a higher rate of heart disease compared to patients without exercise-induced angina?

---

3. **A question about the combination of age AND maximum heart rate (thalch):**

   How does maximum heart rate achieved differ between older and younger patients with heart disease?

---

### 5.2 Identify One Question the Data CANNOT Answer (3 points)

Write one question about **medication effects or long-term prognosis** that this dataset cannot answer, and explain why.

**Question:**

Does cholesterol-lowering medication reduce long-term mortality in patients diagnosed with heart disease?

**Why it cannot be answered with this data:**

The dataset do not contain medication information

No follow-up data

---

### 5.3 Grouping Analysis (3 points)

Answer this question using a groupby analysis:

**"What is the average ST depression (oldpeak) for each type of chest pain (cp)?"**

(Note: You'll need to handle any missing values in oldpeak!)

**Your Code:**

In [53]:

cp_oldpeak = df.groupby("cp")["oldpeak"].mean()

print(cp_oldpeak)


cp
asymptomatic       1.163226
atypical angina    0.304192
non-anginal        0.640000
typical angina     1.070732
Name: oldpeak, dtype: float64


**Your Interpretation:**

Which chest pain type has the highest average ST depression? Which has the lowest? What might this tell us about the relationship between chest pain type and objective measures of cardiac stress?

In [57]:
print("The chest pain type has the highest average ST depression is", cp_oldpeak.idxmax())
print("The chest pain type has the lowest average ST depression is", cp_oldpeak.idxmin())
print("Higher values indicate greater myocardial stress. atients who are asymptomatic show the highest average ST depression. Patients with atypical angina show the lowest. ")

The chest pain type has the highest average ST depression is asymptomatic
The chest pain type has the lowest average ST depression is atypical angina
Higher values indicate greater myocardial stress. atients who are asymptomatic show the highest average ST depression. Patients with atypical angina show the lowest. 


---

## Part 6: Target Variable Analysis (Bonus - 5 points)

The `num` column is our **target variable** - what we're trying to predict. It has 5 possible values (0-4) representing the severity of heart disease narrowing. Analyze its distribution.

**Your Code:**

In [58]:
# Show the count and percentage for each value of num
num_counts = df["num"].value_counts().sort_index()
num_percent = df["num"].value_counts(normalize=True).sort_index() * 100

print("Severity Level Distribution:")
print(pd.DataFrame({
    "Count": num_counts,
    "Percent": num_percent.round(2)
}))

# Also create a binary version (0 = no disease, 1 = disease present) and show its distribution
df["disease_binary"] = np.where(df["num"] > 0, 1, 0)

binary_counts = df["disease_binary"].value_counts()
binary_percent = df["disease_binary"].value_counts(normalize=True) * 100

print("Binary Disease Distribution:")
print(pd.DataFrame({
    "Count": binary_counts,
    "Percent": binary_percent.round(2)
}))


Severity Level Distribution:
     Count  Percent
num                
0      411    44.67
1      265    28.80
2      109    11.85
3      107    11.63
4       28     3.04
Binary Disease Distribution:
                Count  Percent
disease_binary                
1                 509    55.33
0                 411    44.67


### Bonus Questions:

**1. What percentage of patients in this dataset have some form of heart disease (num > 0)?**

Your answer:

---

**2. Is this dataset balanced or imbalanced between disease/no disease?**

Your answer:

---

**3. Looking at the 5 severity levels (0-4), which level is most common? Is this distribution what you'd expect clinically?**

Your answer:

---

**4. How might the multi-severity nature of the target (0-4) versus a simple binary (disease/no disease) affect how you'd approach a machine learning problem?**

Your answer:

---

## Submission Checklist

Before submitting, verify you have completed:

- [ ] **Part 1:** All 10 inspection steps with code AND written findings
- [ ] **Part 2:** Complete data dictionary with all 16 columns filled in
- [ ] **Part 2:** Answered all 4 clinical research questions
- [ ] **Part 3:** All 3 validation checks with code and answers
- [ ] **Part 4:** Created `chol_category` column using **Cholesterol Risk Categories**
- [ ] **Part 4:** Calculated heart disease rate by cholesterol category with interpretation
- [ ] **Part 5:** Three research questions (slope, exang, age+thalch)
- [ ] **Part 5:** One unanswerable question about medication/prognosis
- [ ] **Part 5:** ST depression by chest pain type groupby analysis
- [ ] **Bonus (Optional):** Target variable analysis

---

## Grading Rubric

| Component | Points | Requirements for Full Credit |
|-----------|--------|------------------------------|
| Part 1: 10-Point Inspection | 40 | All 10 steps complete with working code AND thoughtful written analysis |
| Part 2: Data Dictionary | 20 | All 16 columns documented with correct feature types and clinical research |
| Part 3: Data Validation | 15 | All validation checks complete with working code and insightful answers |
| Part 4: Cholesterol Categories | 10 | Working code that creates correct groups (including invalid handling) AND meaningful interpretation |
| Part 5: Research Questions | 15 | Three good questions in specified areas, one clear limitation, groupby analysis complete |
| **Bonus:** Target Analysis | +5 | Thoughtful analysis of target variable distribution with clinical connection |
| **Total** | 100 (+5 bonus) | |

---

## Hints (Read Before You Get Stuck!)

### ⚠️ Common Pitfalls:

1. **Values of 0 that aren't really zero**
   - Some columns (like `trestbps` and `chol`) have 0 values that represent missing data, not actual measurements
   - The `.isnull()` method won't detect these!
   - Always examine your data with `value_counts()` and check if extreme values make clinical sense

2. **Missing data varies by source**
   - The four medical centers have different data completeness
   - Some columns like `ca`, `thal`, and `slope` have extensive missing data
   - Understanding WHY data is missing is as important as knowing that it's missing

3. **The target variable has multiple levels**
   - `num` ranges from 0-4, not just 0-1
   - For many analyses, you'll want to convert this to binary (disease present/absent)

4. **Sex is coded as text**
   - Unlike some versions of this dataset, sex is "Male"/"Female" not 0/1

### 💡 Pro Tips:

- Use `value_counts(dropna=False)` to see if there are any null values
- When something seems weird (like 0 cholesterol), investigate it—don't assume it's valid
- Cross-reference columns: do the patterns make clinical sense together?
- Read the original UCI documentation for additional context
- Always state what the data CAN tell you vs. what would require additional information

---

## Useful Resources

- **UCI Repository Page:** https://archive.ics.uci.edu/ml/datasets/heart+disease
- **Cholesterol Levels Guide:** https://www.heart.org/en/health-topics/cholesterol/about-cholesterol
- **Stress Test Information:** https://www.mayoclinic.org/tests-procedures/stress-test/about/pac-20385234
- **ECG ST Segment:** https://litfl.com/st-segment-ecg-library/
- **Pandas Documentation:** https://pandas.pydata.org/docs/

---

*Remember: "Every Column Tells a Story" - your job is to figure out what that story is!*

---

**Due Date:** [See Canvas]

**Submission:** Upload your completed Jupyter notebook (.ipynb) to Canvas